lib/SILOptimizer/IPO/FunctionSignatureOpts.cpp

```diff
@@ -501,90 +260,28 @@ unsigned ArgumentDescriptor::updateOptimizedBBArgs(SILBuilder &Builder,
 
 namespace {
 
-// Helper class that analyzes the parameters of a function to
-// determine how we can modify the function signature to improve the
-// quality of the code that we generate.
-class SignatureAnalyzer {
-  /// The function that we are analyzing.
-  SILFunction *F;
-
-  RCIdentityFunctionInfo *RCIA;
-
-  AliasAnalysis *AA;
-
-  /// Does any call inside the given function may bind dynamic 'Self' to a
-  /// generic argument of the callee.
-  bool MayBindDynamicSelf;
-
-  /// Did we decide to change the self argument? If so we need to
-  /// change the calling convention 'method' to 'freestanding'.
-  bool ShouldModifySelfArgument = false;
-
-  /// A list of structures which present a "view" of precompiled information on
-  /// an argument that we will use during our optimization.
-  llvm::SmallVector<ArgumentDescriptor, 8> ArgDescList;
-
-  /// Keep a "view" of precompiled information on the direct results
-  /// which we will use during our optimization.
-  llvm::SmallVector<ResultDescriptor, 4> ResultDescList;
-
-  llvm::BumpPtrAllocator &Allocator;
-
-public:
-  SignatureAnalyzer(SILFunction *F, RCIdentityFunctionInfo *RCIA,
-                    AliasAnalysis *AA,
-                    llvm::BumpPtrAllocator &Allocator)
-      : F(F), RCIA(RCIA), AA(AA), MayBindDynamicSelf(computeMayBindDynamicSelf(F)),
-        Allocator(Allocator) {}
-
-  bool analyze();
-  bool analyzeParameters();
-  bool analyzeResult();
-
-  /// Returns the mangled name of the function that should be generated from
-  /// this function analyzer.
-  std::string getOptimizedName() const;
-
-  bool shouldModifySelfArgument() const { return ShouldModifySelfArgument; }
-  ArrayRef<ArgumentDescriptor> getArgDescList() const { return ArgDescList; }
-  MutableArrayRef<ResultDescriptor> getResultDescList() {return ResultDescList;}
-  MutableArrayRef<ArgumentDescriptor> getArgDescList() { return ArgDescList; }
-  SILFunction *getAnalyzedFunction() const { return F; }
-
-private:
-  /// Is the given argument required by the ABI?
-  ///
-  /// Metadata arguments may be required if dynamic Self is bound to any generic
-  /// parameters within this function's call sites.
-  bool isArgumentABIRequired(SILArgument *Arg) {
-    // This implicitly asserts that a function binding dynamic self has a self
-    // metadata argument or object from which self metadata can be obtained.
-    return MayBindDynamicSelf && (F->getSelfMetadataArgument() == Arg);
-  }
-};
-
 /// A class that contains all analysis information we gather about our
 /// function. Also provides utility methods for creating the new empty function.
 class SignatureOptimizer {
-  SignatureAnalyzer &Analyzer;
+  FunctionSignatureFunctionInfo *FSFI;
```

Comment:
Since this is never nullptr, you may consider making it a reference instead, like SignatureAnalyzer was.

Also, I think FunctionSignatureInfo might be a better name than FunctionSignatureFunctionInfo. The latter makes it more clear that it's information for a single function, but I don't think the former is going to be confusing to anyone.